lib/recurly/hosted-fields.js

```diff
@@ -96,17 +101,94 @@ export class HostedFields extends Emitter {
+    let tabbableItems = tabbable(global.document.body);
+
+    // Add hosted fields to the tabbable list
+    this.fields.forEach(field => {
+      if (field.container.compareDocumentPosition(tabbableItems[0]) & Node.DOCUMENT_POSITION_FOLLOWING) {
+        tabbableItems.unshift(field);
+      } else if (field.container.compareDocumentPosition(last(tabbableItems)) & Node.DOCUMENT_POSITION_PRECEDING) {
+        tabbableItems.push(field);
+      } else {
+        tabbableItems.find((item, i) => {
+          let el = item instanceof HostedField ? item.container : item;
+          let nextItem = tabbableItems[i + 1];
+          let nextEl = nextItem instanceof HostedField ? nextItem.container : nextItem;
+          if (!dom.element(el)) return;
+          if (!dom.element(nextEl)) return;
+          if (!(field.container.compareDocumentPosition(el) & Node.DOCUMENT_POSITION_PRECEDING)) return;
+          if (!(field.container.compareDocumentPosition(nextEl) & Node.DOCUMENT_POSITION_FOLLOWING)) return;
+          tabbableItems.splice(i + 1, 0, field);
+          return true;
+        });
+      }
+    });
+
+    // Find the origin within the tabbable list, and focus in our intended direction
+    tabbableItems.find((el, i) => {
+      if (el !== origin) return;
+      let destination;
+      let fieldsSlice;
+
+      if (direction === 'previous') {
+        if (origin instanceof HostedField) {
+          const upperBound = tabbableItems.indexOf(origin);
+          fieldsSlice = tabbableItems.slice(tabbableItems[0], upperBound)
+          for (let i = upperBound; i >= 0; i--) {
+            /* Assume that we are trying to tab to the penultimate
+             * element in the flieldsSlice collection. This covers
+             * the case where we're trying to tab to a previous input
+             * that is not a hosted field
+              */
+            destination = fieldsSlice[upperBound - 1];
+            if (fieldsSlice[i] instanceof HostedField) {
+              destination = fieldsSlice[i];
+              break;
+            }
+          }
+        } else {
+            destination = tabbableItems[i - 1] || last(tabbableItems);
+        }
+      } else {
+          if (origin instanceof HostedField) {
+            fieldsSlice = tabbableItems.slice(tabbableItems.indexOf(origin) + 1, tabbableItems.length - 1);
+            fieldsSlice.find((field, fieldIndex) => {
+              if (field instanceof HostedField) {
+                destination = field;
+                return destination;
+              }
+            });
+          } else {
+              destination = tabbableItems[i + 1] || tabbableItems[0];
+          }
+          if (destination === undefined) {
+            fieldsSlice.find((field, fieldIndex) => {
+              if (!(dom.isHiddenInput(field))) {
+                destination = field;
+                return destination;
+              }
+            });
+          }
+      }
```

Comment:
There's a bit of repetition here, and the intent of the fieldSlice isn't clear. I have to believe the nested looping shouldn't be needed either, since we know our origin and direction within the tabbable set when onTab is called.